src/symbols.c

```diff
@@ -1306,44 +1369,40 @@ static void free_iter_slice_list(gpointer data)
  * previous data is not overwritten if the key is duplicated, but rather the
  * two values are kept in a list
  *
- * table is: GHashTable<TMTag, GList<GList<TMTag>>> */
+ * table is: GHashTable<TMTag, GTree<line_num, GList<GList<TMTag>>>> */
 static void tags_table_insert(GHashTable *table, TMTag *tag, GList *data)
 {
-	GList *list = g_hash_table_lookup(table, tag);
+	GTree *tree = g_hash_table_lookup(table, tag);
+	if (!tree)
+	{
+		tree = g_tree_new_full(tree_cmp, NULL, NULL, NULL);
+		g_hash_table_insert(table, tag, tree);
+	}
+	GList *list = g_tree_lookup(tree, GINT_TO_POINTER(tag->line));
 	list = g_list_prepend(list, data);
-	g_hash_table_insert(table, tag, list);
+	g_tree_insert(tree, GINT_TO_POINTER(tag->line), list);
 }
 
 
-/* looks up the entry in @table that better matches @tag.
- * if there are more than one candidate, the one that has closest line position to @tag is chosen */
+/* looks up the entry in @table that best matches @tag.
+ * if there is more than one candidate, the one that has closest line position to @tag is chosen */
 static GList *tags_table_lookup(GHashTable *table, TMTag *tag)
 {
-	GList *data = NULL;
-	GList *node = g_hash_table_lookup(table, tag);
-	if (node)
-	{
-		glong delta;
-		data = node->data;
-
-#define TAG_DELTA(a, b) ABS((glong) TM_TAG(a)->line - (glong) TM_TAG(b)->line)
+	TreeSearchData user_data = {-1, tag->line, FALSE};
+	GTree *tree = g_hash_table_lookup(table, tag);
 
-		delta = TAG_DELTA(((GList *) node->data)->data, tag);
-		for (node = node->next; node; node = node->next)
-		{
-			glong d = TAG_DELTA(((GList *) node->data)->data, tag);
-
-			if (d < delta)
-			{
-				data = node->data;
-				delta = d;
-			}
-		}
-
-#undef TAG_DELTA
+	if (tree)
+	{
+		GList *list;
 
+		g_tree_search(tree, (GCompareFunc)tree_search_func, &user_data);
+		list = g_tree_lookup(tree, GINT_TO_POINTER(user_data.found_line));
```

Comment:
should this rather be guarded to whether search succeeded?  `if (g_tree_search(tree, ...)) { list = lookup() ... }`